File: src/main/kotlin/io/github/dylmeadows/eontimer/service/CalibrationService.kt
package io.github.dylmeadows.eontimer.service

import io.github.dylmeadows.eontimer.model.settings.TimerSettingsModel
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class CalibrationService @Autowired constructor(
    private val timerSettingsModel: TimerSettingsModel) {

    private val console get() = timerSettingsModel.console

    fun toMillis(delays: Long): Long =
        Math.round(delays * console.frameRate)

    fun toDelays(millis: Long): Long =
        Math.round(millis / console.frameRate)

    fun createCalibration(delay: Long, second: Long): Long =
        toMillis(delay - toDelays(second * 1000))

    fun calibrate(value: Long): Long {
        return if (!timerSettingsModel.precisionCalibrationMode)
            toMillis(value)
        else
            value
    }
}

File: services/CalibrationService.cpp
//
// Created by Dylan Meadows on 2020-03-10.
//

#include "CalibrationService.h"
#include "SettingsService.h"
#include <models/Console.h>
#include <cmath>

namespace service::CalibrationService {
    int toDelays(const int milliseconds) {
        const double framerate = model::getFramerate(Settings::getConsole());
        return static_cast<int>(std::round(milliseconds / framerate));
    }

    int toMilliseconds(const int delays) {
        const double framerate = model::getFramerate(Settings::getConsole());
        return static_cast<int>(std::round(delays * framerate));
    }

    int createCalibration(const int delays, const int seconds) {
        return toMilliseconds(delays - toDelays(seconds * 1000));
    }
}